### PR TITLE
Improve axis tick generation and formatting

### PR DESCRIPTION
**Pull request by pavelgruba**

### CODE REVIEW

File: js/viz/axes/tick_generator.js [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+        calculateTickIntervalDateTime,
+        calculateMinorTickIntervalDateTime,
+        getTickIntervalByCustomTicks(getValue, dateUtils.convertMillisecondsToDateUnits),
+        dateToMilliseconds,
+        calculateTicks(addIntervalDate, options.endOnTick ? floor : ceil),
+        calculateMinorTicks(getValue, addIntervalDate, options.endOnTick ? floor : ceil, addIntervalDate, getValue)
+    );
+}
+
+// {
```
Comment (dxBeardedBear) [inline js/viz/axes/tick_generator.js]: Commented options? For what?

File: js/viz/core/themes/generic.light.js [hunk 2]
```diff
@@ -229,7 +229,8 @@ registerTheme({
     },
     "chart:common:axis": {
         visible: false,
-        setTicksAtUnitBeginning: true,
+        //TODO
+        //setTicksAtUnitBeginning: true,
```
Comment (dxBeardedBear) [inline js/viz/core/themes/generic.light.js]: todo or not todo?

File: js/viz/core/themes/generic.light.js [hunk 3]
```diff
@@ -994,12 +996,14 @@ registerTheme({
...
         scale: {
+            valueMarginsEnabled: true,
             width: 1,
             color: BLACK,
             opacity: 0.1,
             showCustomBoundaryTicks: true,
             showMinorTicks: true, //DEPRECATED IN 15_2
-            setTicksAtUnitBeginning: true,
+            //TODO
+            //setTicksAtUnitBeginning: true,
```
Comment (dxBeardedBear) [inline js/viz/core/themes/generic.light.js]: another one

File: js/viz/range_selector/range_selector.js [hunk 4]
```diff
@@ -496,6 +511,8 @@ function prepareScaleOptions(scaleOption, seriesDataSource, incidentOccurred) {
     }
 
     scaleOption.valueType = valueType;
+    //TODO
```
Comment (dxBeardedBear) [inline js/viz/range_selector/range_selector.js]: TODO what?

File: testing/tests/DevExpress.viz.core/baseAxis.tests.js [hunk 5]
```diff
@@ -1576,3 +912,1106 @@ QUnit.test("Get viewport. Only max is defined", function(assert) {
...
+            min: 1,
+            max: 1000000,
+            minVisible: 1,
+            maxVisible: 1000000
+        },
+        isArgumentAxis: true
+    });
+});
+
+//TODO is it something real?
```
Comment (dxBeardedBear) [inline testing/tests/DevExpress.viz.core/baseAxis.tests.js]: Is this informative todo?

File: js/viz/axes/base_axis.js [hunk 6]
```diff
@@ -37,17 +36,42 @@ var vizUtils = require("../core/utils"),
...
+        axisType: options.type,
+        dataType: options.dataType,
+        logBase: options.logarithmBase,
+
+        axisDivisionFactor: options.axisDivisionFactor || DEFAULT_AXIS_DIVISION_FACTOR,
+        minorAxisDivisionFactor: options.minorAxisDivisionFactor || DEFAULT_MINOR_AXIS_DIVISION_FACTOR,
+        numberMultipliers: options.numberMultipliers,
+        calculateMinors: options.minorTick.visible || options.minorGrid.visible || options.calculateMinors,
+
+        //TODO new options
```
Comment (dxBeardedBear) [inline js/viz/axes/base_axis.js]: Is the goal achieved?

File: js/viz/axes/base_axis.js [hunk 7]
```diff
@@ -1097,44 +1025,202 @@ Axis.prototype = {
...
+            range.addRange({
+                minVisible: minVisible,
+                maxVisible: maxVisible,
+                interval: interval
+            });
         }
+        that._translator.updateBusinessRange(range);
+    },
+
+    //for internal use only
```
Comment (dxBeardedBear) [inline js/viz/axes/base_axis.js]: The underscore at the beginning of the function name indicates 'for internal use only', isn't it?

File: js/viz/chart_components/advanced_chart.js [hunk 8]
```diff
@@ -174,13 +196,6 @@ var AdvancedChart = BaseChart.inherit({
...
-    _drawAxes: function(drawOptions, panesBorderOptions) {
-        this._restoreOriginalBusinessRange();
-        this._prepareAxesAndDraw(drawOptions, panesBorderOptions);
-    },
-
-    _restoreOriginalBusinessRange: _noop,
-
     _appendAdditionalSeriesGroups: function() {
         this._crosshairCursorGroup.linkAppend();
         //this._legendGroup.linkAppend();
```
Comment (dxBeardedBear) [inline js/viz/chart_components/advanced_chart.js:201]: Old commented code

File: js/viz/range_selector/range_selector.js [hunk 9]
```diff
@@ -460,6 +474,7 @@ function updateScaleOptions(scaleOptions, seriesDataSource, translatorRange, tic
         intervals = getIntervalCustomTicks(scaleOptions);
         scaleOptions.customMinorTicks = intervals.altIntervals;
         scaleOptions.customTicks = intervals.intervals;
+        //TODO rework
```
Comment (dxBeardedBear) [inline js/viz/range_selector/range_selector.js]: Uninformative

File: js/viz/core/themes/generic.light.js (hunk 2 above)
Comment (Krijovnick) [inline js/viz/core/themes/generic.light.js]: Perhaps it should be removed

**Review by alexanderPolosatov — APPROVED** [review]

**Review by roman-simionov — APPROVED** [review]

**Pull request merged**